### PR TITLE
Feature: Clone

### LINKED ISSUE #210: CLI: clone/promote
- clone a jail
- promote the cloned jail

### PR DESCRIPTION
**Pull request by gronke**
closes #210

Introduces the `ioc clone` CLI command:
```
ioc clone <FROM_JAIL> <TO_JAIL>
ioc promote <TO_JAIL>
ioc destroy <FROM_JAIL>
```

- Recursive cloning of datasets / snapshots
- Export jails to a dataset
- Jails can be cloned from existing ones
- Cloned jails can be promoted (so that the parent can be deleted)

### CODE REVIEW

File: iocage/lib/ZFS.py [hunk 1]
```diff
@@ -24,16 +24,31 @@
...
 import iocage.lib.errors
 
+Logger = iocage.lib.Logger.Logger
+
 
 class ZFS(libzfs.ZFS):
     """libzfs enhancement module."""
 
-    logger: typing.Optional[iocage.lib.Logger.Logger] = None
+    _logger: typing.Optional['iocage.lib.Logger.Logger']
```
Comment (gronke) [inline iocage/lib/ZFS.py]: This change is required for compliance with the typing. The logger on ZFS is optional and is not set when initializing the object, but can be set at any later time (using the setter method).

**Comment by gronke** [comment]
In an offline discussion @yonk42 mentioned that a parent of a cloned jail cannot be deleted unless they get promoted. This is true and needs to be addressed, but I'm unsure whether automatically promoting could be an option. If we do not automatically promote another jail on deletion of the parent, further user interaction is required.

**Comment by gronke** [comment]
@foo2342 brought up an interesting point while discussing the problem. Instead of automatically promoting the next available sibling the former parent could be moved in an archive dataset (instead of `iocage/jails`), so that the datasets persists until the last clone was deleted without getting recognized as startable jail. It is reasonable to delete every snapshot that was not cloned by another jail on parent jail deletion.

File: iocage/lib/ZFS.py (hunk 1 above)
Comment (igalic) [reply iocage/lib/ZFS.py]: i think it might be time that we (once again) rethink our logger (initialization (method))

Comment (gronke) [reply iocage/lib/ZFS.py]: This is a special occasion. Having a logger in this class available improves our output, but the ZFS class itself should not depend on having a logger. (Imagine you want to use the module in another context than libiocage). I also remember an issue overriding the default ZFS __init__ method.

**Pull request merged**